interceptors/tracing/interceptors_test.go

```diff
@@ -0,0 +1,313 @@
+)
+
+var (
+	id               int64 = 0
+	traceIDHeaderKey       = "traceid"
+	spanIDHeaderKey        = "spanid"
+)
+
+func extractFromContext(ctx context.Context, kind tracing.SpanKind) *mockSpan {
+	var m metadata.MD
+	if kind == tracing.SpanKindClient {
+		m, _ = metadata.FromOutgoingContext(ctx)
+	} else {
+		m, _ = metadata.FromIncomingContext(ctx)
+	}
+
+	traceIDValues := m.Get(traceIDHeaderKey)
+	if len(traceIDValues) == 0 {
+		return nil
+	}
+	spanIDValues := m.Get(spanIDHeaderKey)
+	if len(spanIDValues) == 0 {
+		return nil
+	}
+
+	return &mockSpan{
+		traceID: traceIDValues[0],
+		spanID:  spanIDValues[0],
+	}
+}
+
+func injectWithContext(ctx context.Context, span *mockSpan, kind tracing.SpanKind) context.Context {
+	var m metadata.MD
+	if kind == tracing.SpanKindClient {
+		m, _ = metadata.FromOutgoingContext(ctx)
+	} else {
+		m, _ = metadata.FromIncomingContext(ctx)
+	}
+	m = m.Copy()
+
+	m.Set(traceIDHeaderKey, span.traceID)
+	m.Set(spanIDHeaderKey, span.spanID)
+
+	ctx = metadata.NewOutgoingContext(ctx, m)
+	return ctx
+}
+
+func genID() string {
+	return strconv.FormatInt(atomic.AddInt64(&id, 1), 10)
+}
+
+// Implements Tracker
+type mockTracer struct {
+	spanStore map[string]*mockSpan
+}
+
+func (t *mockTracer) ListSpan(kind tracing.SpanKind) []*mockSpan {
+	var spans []*mockSpan
+	for _, v := range t.spanStore {
+		if v.kind == kind {
+			spans = append(spans, v)
+		}
+	}
+	return spans
+}
+
+func (t *mockTracer) Reset() {
+	t.spanStore = make(map[string]*mockSpan)
+}
+
+func newMockTracer() *mockTracer {
+	return &mockTracer{
+		spanStore: make(map[string]*mockSpan),
+	}
+}
+
+func (t *mockTracer) Start(ctx context.Context, spanName string, kind tracing.SpanKind) (context.Context, tracing.Span) {
+	span := mockSpan{
+		spanID:     genID(),
+		name:       spanName,
+		kind:       kind,
+		statusCode: codes.OK,
+	}
+
+	parentSpan := extractFromContext(ctx, kind)
+	if parentSpan != nil {
+		// Fetch span from context as parent span
+		span.traceID = parentSpan.traceID
+		span.parentSpanID = parentSpan.spanID
+	} else {
+		span.traceID = genID()
+	}
+
+	t.spanStore[span.spanID] = &span
+	if kind == tracing.SpanKindClient {
+		ctx = injectWithContext(ctx, &span, kind)
+	}
+	return ctx, &span
+}
+
+// Implements Span
+type mockSpan struct {
+	traceID      string
+	spanID       string
+	parentSpanID string
+
+	name string
+	kind tracing.SpanKind
+	end  bool
+
+	statusCode    codes.Code
+	statusMessage string
+
+	msgSendCounter     int
+	msgReceivedCounter int
+	eventNameList      []string
+	attributesList     [][]tracing.KeyValue
+}
+
+func (s *mockSpan) SetAttributes(attrs ...tracing.KeyValue) {
+	s.attributesList = append(s.attributesList, attrs)
+}
+
+func (s *mockSpan) End() {
+	s.end = true
+}
+
+func (s *mockSpan) SetStatus(code codes.Code, message string) {
+	s.statusCode = code
+	s.statusMessage = message
+}
+
+func (s *mockSpan) AddEvent(name string, attrs ...tracing.KeyValue) {
```

Comment:
Idea: Remove kv and use `keyvals ...any` as in go-kit logger https://github.com/go-kit/kit/tree/master/log ? (:

Reply:
well `..interface{}`, since we can't just support Go 1.18 yet

Reply:
Then we can use something like 

```
// borrowed from https://github.com/go-logfmt/logfmt/blob/main/encode.go#L75
func kvToAttr(keyvals ...interface{}) []attribute.KeyValue {
	if len(keyvals) == 0 {
		return nil
	}
	if len(keyvals)%2 == 1 {
		keyvals = append(keyvals, nil)
	}

	j := 0
	ret := make([]attribute.KeyValue, len(keyvals)/2)
	for i := 0; i < len(keyvals); i += 2 {
		k, ok := anyToString(keyvals[i])
		if !ok {
			k = "<unsupported key type>"
		}
		
		v, ok := anyToString(keyvals[i+1])
		if !ok {
			v = "<unsupported value type>"
		}
		
		ret[j] = attribute.String(k, v)
		j++
	}
	return ret
}

func anyToString(value interface{}) (string, bool) {
	switch v := value.(type) {
	case nil:
		return "nil", true
	case string:
		return v, true
	case []byte:
		return string(v), true
	case error:
		return v.Error(), true
	case fmt.Stringer:
		return v.String(), true
	default:
		rvalue := reflect.ValueOf(value)
		switch rvalue.Kind() {
		case reflect.Array, reflect.Chan, reflect.Func, reflect.Map, reflect.Slice, reflect.Struct:
			return "", false
		case reflect.Ptr:
			if rvalue.IsNil() {
				return "nil", true
			}
			return anyToString(rvalue.Elem().Interface())
		}
		return fmt.Sprint(v), true
	}
}

```